nixos/etc-overlay: make the etc overlay compatible with nixos-enter and nixos-install

r-vdp opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

When using nixos-enter (and so also nixos-install) on a system with etc-overlay enabled, the activation script gets called directly, and there is no systemd running.
This violates a couple of assumptions in the etc-overlay activation script which assumed that it only ever ran when switching into a new generation and that the very first /etc would always have been set up by the systemd initrd.

As more and more things are being moved into systemd components (initrd services, mount units, tmpfiles, etc), I think that it is going to become increasingly difficult to stay compatible with these tools, but at least for now there is no real alternative and so we probably want to be able to install systems with etc-overlay enabled.

Fixes #364006
Fixes #319533 

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

r-vdp committed:
nixos/etc-overlay: make the etc overlay compatible with nixos-enter and nixos-install

When using nixos-enter (and so also nixos-install) on a system with etc-overlay enabled,
he activation script gets called directly, and there is no systemd running.
This violates a couple of assumptions in the etc-overlay activation script which
assumed that it only ever ran when switching into a new generation and that
the very first /etc would always have been set up by the systemd initrd.

As more and more things are being moved into systemd components (initrd services,
mount units, tmpfiles, etc), I think that it is going to become increasingly
difficult to stay compatible with these tools, but at least for now there is
no real alternative and so we probably want to be able to install systems
with etc-overlay enabled.

r-vdp commented on nixos/modules/system/etc/etc.nix:
```diff
@@ -250,69 +250,83 @@ in
+              # Mount the new /etc overlay to a temporary private mount.
+              # This needs the indirection via a private bind mount because you
+              # cannot move shared mounts.
+              tmpEtcMount=$(TMPDIR="" mktemp --tmpdir=/tmp --directory -t nixos-etc.XXXXXXXXXX)
```
Here and above we force `TMPDIR` to empty to avoid an issue with this variable leaking into the chroot from outside when using nixos-install.
I didn't find a better way to fix this yet.

In any case, further down in this script we currently also assume that `TMPDIR=/tmp`, so we're not losing much currently.

r-vdp replied:
This was changed to use /run in a later commit.

r-vdp commented on nixos/modules/system/etc/etc-activation.nix:
```diff
@@ -15,6 +15,7 @@
       system.activationScripts.etc = lib.stringAfter [
         "users"
         "groups"
+        "specialfs"
```
This is needed so that we have /proc set up, since the activation script for the etc overlay needs to query the kernel for mounts.

Mic92 commented:
> I tried, patch is rejected (unstable?)
> 
> ```
> patching file nixos/modules/system/etc/etc-activation.nix
> Hunk #1 FAILED at 15.
> 1 out of 1 hunk FAILED -- saving rejects to file nixos/modules/system/etc/etc-activation.nix.rej
> patching file nixos/modules/system/etc/etc.nix
> Hunk #1 FAILED at 250.
> Hunk #2 FAILED at 321.
> 2 out of 2 hunks FAILED -- saving rejects to file nixos/modules/system/etc/etc.nix.rej
> patching file pkgs/by-name/ni/nixos-enter/nixos-enter.sh
> ```
> 
> I will try check / check branch etc

There is no merge conflict with master...

r-vdp commented:
Yeah, there's the massive reformat commit that is on master and not yet on nixos-unstable, so that causes the conflict.

phaer reviewed (approved):
Tested this on a hetzner cloud machine, which ran into https://github.com/NixOS/nixpkgs/issues/364006 before. Works as expected with this, thanks a lot! 

Nice catch with the missing /proc. I think we can iteratively improve the situation around TMPDIR and plenty of other yak-shaving opportunities later on, but this at least unblocks nixos-install with etc.overlay! :)

andre4ik3 commented on nixos/modules/system/etc/etc.nix:
```diff
@@ -250,69 +250,83 @@ in
-            # tracking issue: https://github.com/util-linux/util-linux/issues/2604
-            ${pkgs.move-mount-beneath}/bin/move-mount --move --beneath $tmpEtcMount /etc
+            # These don't exist yet when we're called by nixos-enter
+            mkdir --parents /.rw-etc/upper /.rw-etc/work
```
Wouldn't this always create the `/.rw-etc` folder, even when `system.etc.overlay.mutable = false`? I have all of my systems on read-only `/etc` overlay and the folder doesn't exist there at all, maybe a check could be added so it's only created if `system.etc.overlay.mutable = true` to avoid creating unnecessary folders in the root. Also adding it here means the folder is now created in two places, in addition to inside the initrd: https://github.com/NixOS/nixpkgs/blob/a34c4b8e44635241508126bbd07a1f0219487e5a/nixos/modules/system/etc/etc-activation.nix#L122-L137

r-vdp replied:
Good point, I'll have a look at that tomorrow!

r-vdp replied:
So, the ones created in initrd are the same ones, since `/sysroot` is what will become the real root once we pivot from initrd into the final system. So there is no issue there. And `mkdir --parents` does nothing if the directories already exist.

I pushed a commit to only create them when using a mutable overlay, so that they are not created for the immutable case.

andre4ik3 replied:
Yep, just thought I would mention they are created twice just in case of future code changes or such. Thank you for the fix! :)

benys commented:
I had this error:
```
installing the boot loader...
remounting /etc...
mktemp: failed to create directory via template ‘/tmp/nixos-etc-metadata.XXXXXXXXXX’: No such file or directory
mount: /nix/store/wfk93a6hd6p7rnhyykiyz6qpm3z6mr0a-etc-metadata.erofs: can't find in /etc/fstab.
mount: /etc: wrong fs type, bad option, bad superblock on overlay, missing codepage or helper program, or other error.
       dmesg(1) may have more information after failed mount system call.
Activation script snippet 'etc' failed (32)
warning: /root/.nix-defexpr/channels exists, but channels have been disabled.
warning: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled.
Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset.
Delete the above directory or director
```

andre4ik3 commented:
@benys I think it's a bug in `nixos-enter`, `/tmp` gets created after activation scripts are run, not beforehand (and even the comments in the code say "create `/tmp` for activation script"... but runs after the activation scripts). I fixed it like this:

```diff
diff --git a/pkgs/by-name/ni/nixos-enter/nixos-enter.sh b/pkgs/by-name/ni/nixos-enter/nixos-enter.sh
index 214ab7dfbda4..ae2a9382d0a2 100755
--- a/pkgs/by-name/ni/nixos-enter/nixos-enter.sh
+++ b/pkgs/by-name/ni/nixos-enter/nixos-enter.sh
@@ -99,12 +99,12 @@ chroot_add_resolv_conf "$mountPoint" || echo "$0: failed to set up resolv.conf"
         exec 2>/dev/null
     fi
 
-    # Run the activation script. Set $LOCALE_ARCHIVE to suppress some Perl locale warnings.
-    LOCALE_ARCHIVE="$system/sw/lib/locale/locale-archive" IN_NIXOS_ENTER=1 chroot "$mountPoint" "$system/activate" 1>&2 || true
-
     # Create /tmp. This is needed for nix-build and the NixOS activation script to work.
     # Hide the unhelpful "failed to replace specifiers" errors caused by missing /etc/machine-id.
     chroot "$mountPoint" "$system/sw/bin/systemd-tmpfiles" --create --remove -E 2> /dev/null || true
+
+    # Run the activation script. Set $LOCALE_ARCHIVE to suppress some Perl locale warnings.
+    LOCALE_ARCHIVE="$system/sw/lib/locale/locale-archive" IN_NIXOS_ENTER=1 chroot "$mountPoint" "$system/activate" 1>&2 || true
 )
 
 unset TMPDIR
```

Maybe with this, the `TMPDIR=` overrides aren't necessary?

r-vdp commented:
I wonder why we create this stuff in /tmp to begin with, maybe we should just use /run, which feels more appropriate.

andre4ik3 commented:
The temporary mountpoint is removed after it's moved to `/etc`, and metadata is mounted in `/run`, but there seems to be two different metadata mountpoints. `/run/etc-metadata` is mounted in initrd and `/tmp/nixos-etc-metadata.XXXXXXXXX` in activation script (it seems at least one old version sticks around after rebuild) but I guess this can be cleaned up in a separate PR? 🙃

```
❯ mount | grep etc-metadata
/nix/store/5ffk5rlb5x4m5blnq8b5sska3c5qdjsm-etc-metadata.erofs on /run/etc-metadata type erofs (ro,relatime,user_xattr,acl,cache_strategy=readaround)
/nix/store/zcsbh0zimdxrbfxbr5b35a1jnggqgb5l-etc-metadata.erofs on /tmp/nixos-etc-metadata.Q4ITkDq5QS type erofs (ro,relatime,user_xattr,acl,cache_strategy=readaround)
overlay on /etc type overlay (ro,relatime,lowerdir=/tmp/nixos-etc-metadata.Q4ITkDq5QS::/nix/store/sxjzfbwswzad80254a5ayn0srq51jd1d-etc-lowerdir,redirect_dir=on,metacopy=on)

❯ ls -d /tmp/nixos-etc-metadata*
/tmp/nixos-etc-metadata.oDlDEOAVB7  /tmp/nixos-etc-metadata.Q4ITkDq5QS
```

r-vdp committed:
nixos/etc-overlay: always create the metadata mountpoints in /run

This avoids the dependence on the presence of /tmp, which causes issues
with nixos-install

r-vdp commented:
I added a commit to always create the metadata mountpoints in /run. Could you give this another try?

benys commented:
I looks now like this:
```
remounting /etc...
warning: /root/.nix-defexpr/channels exists, but channels have been disabled.
warning: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled.
Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset.
Delete the above directory or directories to prevent this.
Copied "/nix/store/3abwqv1a1bdycmgaydzfw3a0qzxwk8am-systemd-256.8/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/systemd/systemd-bootx64.efi".
Copied "/nix/store/3abwqv1a1bdycmgaydzfw3a0qzxwk8am-systemd-256.8/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/BOOT/BOOTX64.EFI".
Random seed file /boot/loader/random-seed successfully refreshed (32 bytes).
Created EFI boot entry "Linux Boot Manager".
warning: the group 'nixbld' specified in 'build-users-group' does not exist
remounting /etc...
warning: /root/.nix-defexpr/channels exists, but channels have been disabled.
warning: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled.
Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset.
Delete the above directory or directories to prevent this.
remounting /etc...
warning: /root/.nix-defexpr/channels exists, but channels have been disabled.
warning: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled.
Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset.
Delete the above directory or directories to prevent this.
setting root password...
passwd: Cannot determine your user name.
Setting a root password failed with the above printed error.
You can set the root password manually by executing `nixos-enter --root '/mnt'` and then running `passwd` in the shell of the new system.
```

but now I started install from empty disk, so I will update you

benys commented:
I look good:
```
installing the boot loader...
remounting /etc...
warning: /root/.nix-defexpr/channels exists, but channels have been disabled.
warning: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled.
Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset.
Delete the above directory or directories to prevent this.
Initializing machine ID from random generator.
Created "/boot/EFI".
Created "/boot/EFI/systemd".
Created "/boot/EFI/BOOT".
Created "/boot/loader".
Created "/boot/loader/entries".
Created "/boot/EFI/Linux".
Copied "/nix/store/3abwqv1a1bdycmgaydzfw3a0qzxwk8am-systemd-256.8/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/systemd/systemd-bootx64.efi".
Copied "/nix/store/3abwqv1a1bdycmgaydzfw3a0qzxwk8am-systemd-256.8/lib/systemd/boot/efi/systemd-bootx64.efi" to "/boot/EFI/BOOT/BOOTX64.EFI".
Random seed file /boot/loader/random-seed successfully written (32 bytes).
Created EFI boot entry "Linux Boot Manager".
warning: the group 'nixbld' specified in 'build-users-group' does not exist
remounting /etc...
warning: /root/.nix-defexpr/channels exists, but channels have been disabled.
warning: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled.
Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset.
Delete the above directory or directories to prevent this.
remounting /etc...
warning: /root/.nix-defexpr/channels exists, but channels have been disabled.
warning: /nix/var/nix/profiles/per-user/root/channels exists, but channels have been disabled.
Due to https://github.com/NixOS/nix/issues/9574, Nix may still use these channels when NIX_PATH is unset.
Delete the above directory or directories to prevent this.
setting root password...
passwd: Cannot determine your user name.
Setting a root password failed with the above printed error.
You can set the root password manually by executing `nixos-enter --root '/mnt'` and then running `passwd` in the shell of the new system.
```

r-vdp commented:
It seems that there's some issue with /etc/passwd still.

benys commented:
Yes! this computer was at my office, I restarted it and I cannot login.
I will try investigate...

benys commented:
It works, it doesn't set my password because it doesn't have access to secrets.
Thank you!

r-vdp commented:
> It works, it doesn't set my password because it doesn't have access to secrets. Thank you!

Ok, so /etc/passwd is actually correctly put in place? It's a weird error, "cannot determine your user name", it sounds as if there is no entry in /etc/passwd for the uid.

benys commented:
@r-vdp:
1) I didn't have a "Secret enabled" configuration for new machine. Now it works ok....
2)  I use your repo and I discovered this folders in root:
```
󰈺 ➜ ls -la /
drwxr-xr-x - root 16 gru 18:56 .rw-etc
drwxr-xr-x - root 16 gru 18:56 bin
drwx------ - root  1 sty  1970 boot
drwxr-xr-x - root 16 gru 18:56 dev
drwxr-xr-x - root 16 gru 18:56 etc
drwxr-xr-x - root 16 gru 18:56 home
drwxr-xr-x - root 16 gru 18:56 lib
drwxr-xr-x - root 16 gru 18:56 lib64
drwxr-xr-x - root 27 kwi 16:55 nix
drwx--x--x - root 16 gru 19:15 opt
dr-xr-xr-x - root 16 gru 18:56 proc
drwxr-x--- - root 16 gru 18:56 root
drwxr-xr-x - root 16 gru 19:15 run
drwxr-xr-x - root 16 gru 18:56 srv
dr-xr-xr-x - root 16 gru 19:21 sys
drwxrwxrwt - root 16 gru 19:20 tmp
drwxr-xr-x - root 16 gru 18:56 usr
drwxr-xr-x - root 16 gru 18:56 var
drwxr-xr-x - root 16 gru 18:56 vol
```
It is normal to see .rw-etc folder?

r-vdp commented:
> It is normal to see .rw-etc folder?

Yeah, that directory holds the workdir and the mutable upper layer of the overlayfs that's mounted on /etc.

r-vdp commented:
@nikstur @ElvishJerricco any objections to these changes? I've been using this on 4 systems for the last two weeks and others confirmed that this indeed solves the issues that they were having with nixos-install.

phaer commented:
If there are no objections, it might be okay to merge? New installs of systems with etc.overlay are still broken and if I am reading this thread correctly there's been a few testers and no known open regression?
